File: task-2/src/main/java/ru/biluta/agents/BookBuyerAgent.java
package ru.biluta.agents;

import jade.core.AID;
import jade.core.Agent;
import jade.core.behaviours.Behaviour;
import jade.core.behaviours.TickerBehaviour;
import jade.domain.FIPAAgentManagement.DFAgentDescription;
import jade.domain.FIPAAgentManagement.ServiceDescription;
import jade.domain.FIPAException;
import jade.domain.DFService;
import jade.lang.acl.ACLMessage;
import jade.lang.acl.MessageTemplate;

public class BookBuyerAgent extends Agent {
    private String targetBookTitle;
    private AID bestSeller;
    private int bestPrice = Integer.MAX_VALUE;
    private int repliesCnt = 0; // Счетчик ответов

    protected void setup() {
        Object[] args = getArguments();
        if (args != null && args.length > 0) {
            targetBookTitle = (String) args[0];
            addBehaviour(new TickerBehaviour(this, 10000) { // Периодическое поведение
                protected void onTick() {
                    System.out.println("Trying to buy " + targetBookTitle);
                    // Обновление списка продавцов
                    DFAgentDescription template = new DFAgentDescription();
                    ServiceDescription sd = new ServiceDescription();
                    sd.setType("book-selling");
                    template.addServices(sd);

                    try {
                        DFAgentDescription[] result = DFService.search(myAgent, template);
                        System.out.println("Found the following seller agents:");
                        for (int i = 0; i < result.length; ++i) {
                            System.out.println(result[i].getName().getName());
                        }
                        myAgent.addBehaviour(new RequestPerformer(result));
                    }
                    catch (FIPAException fe) {
                        fe.printStackTrace();
                    }
                }
            });
        } else {
            System.out.println("No target book title specified");
            doDelete();
        }
    }

    private class RequestPerformer extends Behaviour {
        private MessageTemplate mt; // Шаблон для получения ответов
        private int step = 0;
        private DFAgentDescription[] sellerAgents;

        public RequestPerformer(DFAgentDescription[] sellers) {
            this.sellerAgents = sellers;
        }

        public void action() {
            switch (step) {
                case 0:
                    // Send the cfp to all sellers
                    ACLMessage cfp = new ACLMessage(ACLMessage.CFP);
                    for (DFAgentDescription seller : sellerAgents) {
                        cfp.addReceiver(seller.getName());
                    }
                    cfp.setContent(targetBookTitle);
                    cfp.setConversationId("book-trade");
                    cfp.setReplyWith("cfp" + System.currentTimeMillis()); // Unique value
                    myAgent.send(cfp);
                    // Prepare the template to get proposals
                    mt = MessageTemplate.and(MessageTemplate.MatchConversationId("book-trade"),
                            MessageTemplate.MatchInReplyTo(cfp.getReplyWith()));
                    step = 1;
                    break;
                case 1:
                    // Receive all proposals/refusals from seller agents
                    ACLMessage reply = myAgent.receive(mt);
                    if (reply != null) {
                        // Reply received
                        if (reply.getPerformative() == ACLMessage.PROPOSE) {
                            // This is an offer
                            int price = Integer.parseInt(reply.getContent());
                            if (bestSeller == null || price < bestPrice) {
                                // This is the best offer at present
                                bestPrice = price;
                                bestSeller = reply.getSender();
                            }
                        }
                        repliesCnt++;
                        if (repliesCnt >= sellerAgents.length) {
                            // We received all replies
                            step = 2;
                        }
                    } else {
                        block();
                    }
                    break;
                case 2:
                    // Send the purchase order to the seller that provided the best offer
                    ACLMessage order = new ACLMessage(ACLMessage.ACCEPT_PROPOSAL);
                    order.addReceiver(bestSeller);
                    order.setContent(targetBookTitle);
                    order.setConversationId("book-trade");
                    order.setReplyWith("order" + System.currentTimeMillis());
                    myAgent.send(order);
                    // Prepare the template to get the purchase order reply
                    mt = MessageTemplate.and(MessageTemplate.MatchConversationId("book-trade"),
                            MessageTemplate.MatchInReplyTo(order.getReplyWith()));
                    step = 3;
                    break;
                case 3:
                    // Receive the purchase order reply
                    reply = myAgent.receive(mt);
                    if (reply != null) {
                        if (reply.getPerformative() == ACLMessage.INFORM) {
                            // Purchase successful
                            System.out.println(targetBookTitle + " successfully purchased from agent " + reply.getSender().getName());
                            System.out.println("Price = " + bestPrice);
                            myAgent.doDelete();
                        } else {
                            System.out.println("Attempt failed: " + targetBookTitle + " already sold.");
                        }
                        step = 4;
                    } else {
                        block();
                    }
                    break;
            }
        }

        public boolean done() {
            return (step == 4);
        }
    }
}

File: task-2-kotlin/src/main/kotlin/agents/BookBuyerAgent.kt
package agents

import jade.core.Agent
import jade.core.behaviours.CyclicBehaviour
import jade.core.behaviours.OneShotBehaviour
import jade.core.behaviours.TickerBehaviour
import jade.domain.DFService
import jade.domain.FIPAAgentManagement.DFAgentDescription
import jade.domain.FIPAAgentManagement.ServiceDescription
import jade.lang.acl.ACLMessage

class BookBuyerAgent : Agent() {
    private val targetBookTitle = "Книга1"
    private val sellerAgents = mutableListOf<AID>()

    override fun setup() {
        println("Покупатель ${aid.name} готов.")

        // Добавление поведения для поиска продавцов через определенные интервалы времени
        addBehaviour(object : TickerBehaviour(this, 10000) {
            override fun onTick() {
                println("Попытка купить $targetBookTitle")
                // Поиск продавцов
                val template = DFAgentDescription()
                val sd = ServiceDescription()
                sd.type = "book-selling"
                template.addServices(sd)

                try {
                    val result = DFService.search(myAgent, template)
                    sellerAgents.clear()
                    for (dfAgent in result) {
                        sellerAgents.add(dfAgent.name)
                    }
                    println("Найдено продавцов книг: ${sellerAgents.size}")

                    if (sellerAgents.isNotEmpty()) {
                        myAgent.addBehaviour(RequestPerformer())
                    }
                } catch (fe: Exception) {
                    fe.printStackTrace()
                }
            }
        })
    }

    // Поведение для запроса и обработки предложений от продавцов
    inner class RequestPerformer : Behaviour() {
        private var bestSeller: AID? = null
        private var bestPrice: Int = Integer.MAX_VALUE
        private var repliesCnt = 0
        private var step = 0

        override fun action() {
            when (step) {
                0 -> {
                    // Отправка запросов на предложение (CFP) всем продавцам
                    val cfp = ACLMessage(ACLMessage.CFP)
                    for (seller in sellerAgents) {
                        cfp.addReceiver(seller)
                    }
                    cfp.content = targetBookTitle
                    cfp.conversationId = "book-trade"
                    cfp.replyWith = "cfp${System.currentTimeMillis()}"
                    myAgent.send(cfp)
                    step = 1
                }
                1 -> {
                    // Обработка всех ответов (PROPOSE, REFUSE) от продавцов
                    val replies = myAgent.receive()
                    if (replies != null) {
                        repliesCnt++
                        if (replies.performative == ACLMessage.PROPOSE) {
                            val price = replies.content.toInt()
                            if (price < bestPrice) {
                                bestPrice = price
                                bestSeller = replies.sender
                            }
                        }
                        if (repliesCnt >= sellerAgents.size) {
                            step = 2
                        }
                    } else {
                        block()
                    }
                }
                2 -> {
                    // Отправка заказа на покупку (ACCEPT_PROPOSAL) лучшему продавцу
                    val order = ACLMessage(ACLMessage.ACCEPT_PROPOSAL)
                    order.addReceiver(bestSeller)
                    order.content = targetBookTitle
                    order.conversationId = "book-trade"
                    order.replyWith = "order${System.currentTimeMillis()}"
                    myAgent.send(order)
                    step = 3
                }
                3 -> {
                    // Обработка подтверждения покупки (INFORM) от продавца
                    val reply = myAgent.receive()
                    if (reply != null) {
                        if (reply.performative == ACLMessage.INFORM) {
                            println("${targetBookTitle} успешно куплена у агента ${reply.sender.name}")
                            myAgent.doDelete()
                        }
                    } else {
                        block()
                    }
                }
            }
        }

        override fun done(): Boolean {
            return (step == 3 && bestSeller == null)
        }
    }
}
